### PR TITLE
🐛 prevent duplicated files in cloudinit for cloudinit.NewNode

### PR DESCRIPTION
**Pull request by chrischdi**
**What this PR does / why we need it**:

When adding worker nodes the `files` inside cloudinit added each file in `AddtionalFiles` twice.
* by calling `input.prepare` in https://github.com/kubernetes-sigs/cluster-api/blob/master/bootstrap/kubeadm/internal/cloudinit/node.go#L52
  * which added the files in `https://github.com/kubernetes-sigs/cluster-api/blob/master/bootstrap/kubeadm/internal/cloudinit/cloudinit.go#L62`
* and second at https://github.com/kubernetes-sigs/cluster-api/blob/master/bootstrap/kubeadm/internal/cloudinit/node.go#L56

The duplicated files are an issue if the maximum size for cloudinit is reached afterwards, when creating a machine.

I propose removing the duplicate in `bootstrap/kubeadm/internal/cloudinit/node.go`'s `NewNode`, because joining a control-plane node also adds the files using the `prepare` (https://github.com/kubernetes-sigs/cluster-api/blob/master/bootstrap/kubeadm/internal/cloudinit/controlplane_join.go#L61)

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

<sup>
Christian Schlotter <christian.schlotter@daimler.com>, Daimler TSS GmbH, 
<a href="https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md">Imprint</a>
</sup>

### CODE REVIEW

**Comment by chrischdi** [comment]
Sorry for directly creating a Pull Request, I'd also be happy to move over to an issue if this needs further discussion :-)

**Comment by vincepri** [comment]
Can we add a failing test?

**Comment by fabriziopandini** [comment]
+1 to having some test for the change (and to open an issue to add more test on this part of the codebase...)
/ok-to-test

**Comment by sbueringer** [comment]
lgtm for the fix, but yeah a test would be nice. I think it's fine to directly open a PR for fixes like this as the bug is pretty obvious.

**Comment by chrischdi** [comment]
Sure :-), I added a unit test for now.

I thought about using a table test and a `check` function which could vary over different cases could be good to later on add other tests too because the duplicate files case is pretty special.

What do you think?

It would be possible to do the same tests for:
* `NewInitControlPlane` https://github.com/kubernetes-sigs/cluster-api/blob/master/bootstrap/kubeadm/internal/cloudinit/controlplane_init.go#L60
* `NewJoinControlPlane` https://github.com/kubernetes-sigs/cluster-api/blob/9cd9d7c7a51d7d30d3dec18007a109d77c71ef64/bootstrap/kubeadm/internal/cloudinit/controlplane_join.go#L57

**Comment by chrischdi** [comment]
> and to open an issue to add more test on this part of the codebase...

@fabriziopandini I created an issue to increase unit test coverage for the whole package :-) #4634

File: bootstrap/kubeadm/internal/cloudinit/node_test.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+				return
+			}
+			if err := tt.check(got); err != nil {
+				t.Errorf("%v: got = %s", err, got)
+			}
+		})
+	}
+}
+
+func checkDuplicateWriteFiles(b []byte) error {
```
Comment (fabriziopandini) [inline bootstrap/kubeadm/internal/cloudinit/node_test.go:66]: I know this is a little bit of a stretch, but what about making this a little bit more generic by checking that a given list of files is included in cloud init/write_files - without duplicates -, so we can later reuse this check for init/join control plane cloud init unit tests?

something like `checkWriteFiles(files string...) func (b []byte) error { ... }`

File: bootstrap/kubeadm/internal/cloudinit/node_test.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2019 The Kubernetes Authors.
```
Comment (enxebre) [inline bootstrap/kubeadm/internal/cloudinit/node_test.go:2]: 2021

File: bootstrap/kubeadm/internal/cloudinit/node_test.go (hunk 1 above)
Comment (chrischdi) [reply bootstrap/kubeadm/internal/cloudinit/node_test.go:66]: @fabriziopandini I assume you did mean to 
* check all elements in cloudinitData.WriteFiles for duplicates 
* and check for existence of the given files

Is this correct?
I think it never makes sense to have any duplicate right?

Comment (chrischdi) [reply bootstrap/kubeadm/internal/cloudinit/node_test.go:66]: Or according the function name `checkWriteFiles` I think we only should check the given files for duplicates?

Comment (chrischdi) [reply bootstrap/kubeadm/internal/cloudinit/node_test.go]: I adjusted the implementation. PTAL :slightly_smiling_face:

File: bootstrap/kubeadm/internal/cloudinit/node_test.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+				return fmt.Errorf("duplicate file in cloudinit for path %s on index %d in write_files", f.Path, i)
+			}
+			gotFiles[f.Path] = true
+		}
+		for _, file := range files {
+			if !gotFiles[file] {
+				return fmt.Errorf("expected %q to exist in write_files", file)
+			}
+		}
+
```
Comment (fabriziopandini) [inline bootstrap/kubeadm/internal/cloudinit/node_test.go:102]: We should check also that gotFiles and files have the same length

File: bootstrap/kubeadm/internal/cloudinit/node_test.go (hunk 1 above)
Comment (fabriziopandini) [reply bootstrap/kubeadm/internal/cloudinit/node_test.go]: a small nit, but yay!

**Comment by fabriziopandini** [comment]
just a small nit otherwise ready to lgtm pending squash

File: bootstrap/kubeadm/internal/cloudinit/node_test.go (hunk 3 above)
Comment (chrischdi) [reply bootstrap/kubeadm/internal/cloudinit/node_test.go:99]: I also added the check and squashed the commits :-)

File: bootstrap/kubeadm/internal/cloudinit/node_test.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+					},
+				},
+			},
+			checkWriteFiles("/etc/foo.conf", "/run/kubeadm/kubeadm-join-config.yaml", "/run/cluster-api/placeholder"),
+			false,
+		},
+		{
+			"check for existens of /run/kubeadm/kubeadm-join-config.yaml and /run/cluster-api/placeholder",
+			&NodeInput{},
+			checkWriteFiles("/run/kubeadm/kubeadm-join-config.yaml", "/run/cluster-api/placeholder"),
```
Comment (sbueringer) [inline bootstrap/kubeadm/internal/cloudinit/node_test.go:54]: I would prefer adding the files here directly and inlining the check code in l.65+.

It's imho far more common and easier to understand. If I got it correctly they are something like `wantFiles` or `expectedFiles`

Comment (sbueringer) [reply bootstrap/kubeadm/internal/cloudinit/node_test.go:54]: + I think it easier to read if the fields are explicitly added in the struct literal

Comment (chrischdi) [reply bootstrap/kubeadm/internal/cloudinit/node_test.go:54]: The idea behind the test func for me was:
Checking for duplicate Files is a single case and the NewNode function may have others aspects to Test. These other aspects could geht implemented by other functions. Also Not inlining it makes it also usable for the join controlplane equivalent func.

But I'm also OK with inlining it.

Comment (sbueringer) [reply bootstrap/kubeadm/internal/cloudinit/node_test.go:54]: Ah got it. Now it makes sense to me. Fine to keep as is for me.

File: bootstrap/kubeadm/internal/cloudinit/node_test.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+							Permissions: "0644",
+						},
+					},
+				},
+			},
+			checkWriteFiles("/etc/foo.conf", "/run/kubeadm/kubeadm-join-config.yaml", "/run/cluster-api/placeholder"),
+			false,
+		},
+		{
+			"check for existens of /run/kubeadm/kubeadm-join-config.yaml and /run/cluster-api/placeholder",
```
Comment (sbueringer) [inline bootstrap/kubeadm/internal/cloudinit/node_test.go:52]:
```suggestion
			"check for existence of /run/kubeadm/kubeadm-join-config.yaml and /run/cluster-api/placeholder",
```

File: bootstrap/kubeadm/internal/cloudinit/node_test.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+			if err := tt.check(got); err != nil {
+				t.Errorf("%v: got = %s", err, got)
+			}
+		})
+	}
+}
+
+func checkWriteFiles(files ...string) func(b []byte) error {
+	return func(b []byte) error {
+		var cloudinitData struct {
```
Comment (sbueringer) [inline bootstrap/kubeadm/internal/cloudinit/node_test.go:74]: nit: Would it make sense to use the struct we're using when marshaling to parse it here again?

File: bootstrap/kubeadm/internal/cloudinit/node_test.go [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+				return fmt.Errorf("duplicate file in cloudinit for path %s on index %d in write_files", f.Path, i)
+			}
+			gotFiles[f.Path] = true
+		}
+		for _, file := range files {
+			if !gotFiles[file] {
+				return fmt.Errorf("expected %q to exist in write_files", file)
+			}
+		}
+		if len(files) != len(cloudinitData.WriteFiles) {
```
Comment (sbueringer) [inline bootstrap/kubeadm/internal/cloudinit/node_test.go:96]: Can we remove l.86-l.88 when we move this up directly below the unmarshaling?

Would make it possible to skip one check and avoid the implicit reference to a bug nobody will remember when looking at l.87 in a few months

File: bootstrap/kubeadm/internal/cloudinit/node_test.go (hunk 6 above)
Comment (chrischdi) [reply bootstrap/kubeadm/internal/cloudinit/node_test.go:74]: There's no such struct thing. https://github.com/kubernetes-sigs/cluster-api/blob/master/bootstrap/kubeadm/internal/cloudinit/cloudinit.go#L76

`generate()` uses go templates.

**Comment by chrischdi** [comment]
I've added another commit, @sbueringer could you please take another look? I would then squash again afterwards :slightly_smiling_face:

File: bootstrap/kubeadm/internal/cloudinit/node_test.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+		for _, f := range cloudinitData.WriteFiles {
+			gotFiles[f.Path] = true
+		}
+		for _, file := range files {
+			if !gotFiles[file] {
+				return fmt.Errorf("expected %q to exist in write_files", file)
+			}
+		}
+		if len(files) != len(cloudinitData.WriteFiles) {
+			return fmt.Errorf("expected to have %d files generated to WriteFiles, got %d", len(files), len(cloudinitData.WriteFiles))
```
Comment (sbueringer) [inline bootstrap/kubeadm/internal/cloudinit/node_test.go:94]:
```suggestion
			return fmt.Errorf("expected to have %d files generated to write_files, got %d", len(files), len(cloudinitData.WriteFiles))
```

(see l.90)

**Review by sbueringer — COMMENTED** [review]
@chrischdi last nit. Just squash it directly with the fix and I'll lgtm

**Comment by chrischdi** [comment]
Thanks for the review :-) also added the suggestions from Vince and squashed the commits :+1:
